### ISSUE TITLE
Switch the default mockmaker to the inline mockmaker on JDK 17+

### ISSUE THREAD

**Issue by TimvdLippe** [mockito/mockito#2589]
All right, we have been talking about this for a while so let's consolidate discussions. TLDR: more and more use cases are broken (by default) with Mockito and JDK 17. That's because the subclass mockmaker runs into fundamental limitations on JDK 17, but the inline mockmaker works as expected.

There are a lot of issues related to this:
- https://github.com/mockito/mockito/issues/2587
- https://github.com/mockito/mockito/issues/2560
- https://github.com/mockito/mockito/issues/2573
- https://github.com/mockito/mockito/issues/2561 (suspected)
- https://github.com/mockito/mockito/issues/2521 (suspected)
- https://github.com/mockito/mockito/issues/2449 (suspected)

At this point, the rate of issues created is increasing, as more and more of our users are adopting JDK 17+

We have historically talked about this as well: https://github.com/mockito/mockito/issues/1728

When chatting with @raphw about this issue, there were a couple of options that we have available:
1. Keep on using the subclass mockmaker by default on all JDK's and suggest users to use `mockito-inline` whenever appropriate
2. Switch the default to the inline mockmaker for all JDK's and put the subclass mockmaker in an artifact such as `mockito-subclass` (legacy)
3. Detect the current JDK version and make the inline mockmaker the default on JDK 17+, but keep the subclass mockmaker the default for older JDK versions

In my opinion, we should opt for least amount of surprise. E.g. Mockito should "just" work out-of-the-box for the majority of use cases. Therefore, if Mockito (by default) does not work nicely with JDK 17 in more and more cases, I consider that a bug.

That said, I don't think it is a good choice to force all existing users that are on JDK 16 and below to update their codebase wholesale to the new mockmaker. Inheritently, the new mockmaker works slightly different and forcing the update onto users doesn't feel like a nice solution. In the end, it works just fine on JDK 16 on below, so why change it?

Therefore, my proposed way forward is:
1. We publish Mockito 5 which switches the default mockmaker to the inline mockmaker on JDK 17+
2. We update the JavaDoc of the subclass mockmaker to make users aware it is only intended to be used for specific use cases

In terms of timeline, I think we should collective user feedback from the community first. This issue seems like the best place to do so. Once we are confident that we would be able to switch the default, we could publish Mockito 5. For now, what if we take June 2022 as a temporary date to make such a change, unless we receive strong/valid user feedback that should block such a change?

CC @mockito/developers

**Comment by raphw** [comment]
Fully agree with it but with (2). I think we should keep it as it can function well if the JVM is setup correctly. I am for example currently working with some adjustments that will make Mockito work on Graal AOT. This will only work with the subclass mock maker.

**Comment by TimvdLippe** [comment]
Do I understand correctly you would prefer to switch the default for all JDK's, but ship both mockmakers as they target separate use cases? E.g. we don't deprecate the subclass mockmaker, but we maintain the two synonymous and we make the default experience inline, but we allow the subclass mockmaker for JVMs such as Graal.

**Comment by bric3** [comment]
I am personally in favor of option (1).
In my opinion as long as there are use cases for the subclassing strategy (eg android), it shouldn't be abandoned. Even ignoring GraalVM native or Android, I don't think dropping the subclass strategy would be appropriate as long as Mockito base line is not JDK 17.

**Comment by bric3** [comment]
> @bric3 would plan 2 be appropriate for you, with the given addition that we do NOT drop the subclass mockmaker? E.g. we switch the default on JDK 17+, but we keep vendoring both mockmakers as standalone packages so that users who need can choose the most appropriate one.

Yes, I think it's the sensible choice.

@raphw Is this one of the relevant PR : https://github.com/openjdk/jdk/pull/3546

**Comment by raphw** [comment]
Yes, that is what I wanted to suggest. Let's keep the mock makers in one artifact, but let us also have a smart default choice in case that no explicit mock maker is chosen.

I think this is the right thing: default to inline mock maker in Java 17+, subclass mock maker in Java 16-. And also the subclass mock maker if a Graal AOT image code is found. I currently am working with Oracle to adjust Byte Buddy and Mockito to work with Graal AOT. I hope to merge and publish this in a week or two.

**Comment by TimvdLippe** [comment]
Cool! I think we are all in agreement then on the proposed way forward. I will write up a wiki page for users that we can point to, so that we can collect feedback and ask users to test out `mockito-inline`.

**Comment by JojOatXGME** [comment]
I think it would also be nice if you can disable the *inline mock maker* for specific mocks. I think it is good if users can switch to the inline mock maker as early as possible. However, sometimes existing code does not work due to some strange cases. For example I encountered one case where the code did only work because final methods (including `equals` and `hashCode`) were not mocked. Allowing users to enable the inline mock maker by default but keeping the old one for some time could simplify the migration.

**Comment by TimvdLippe** [comment]
Hm, that's a fair point, but I think it is going to be difficult for us, given that mockmakers are globally applied: https://github.com/mockito/mockito/blob/737f0a3517256af6ade37fd0e776a2bb230de553/src/main/java/org/mockito/internal/util/MockUtil.java#L28

We could indeed update the `MockCreationSettings` to add a new field that allows users to supply a `MockMaker` and we use that instead of the globally defined one (here https://github.com/mockito/mockito/blob/737f0a3517256af6ade37fd0e776a2bb230de553/src/main/java/org/mockito/internal/util/MockUtil.java#L44). @JojOatXGME can you file that as a separate feature request and, if possible, send us a WIP PR that would show whether it is feasible to do so?

**Comment by JojOatXGME** [comment]
I created #2626 about that topic. I am not sure if I find time to create a PR within the next few weeks, but lets see.

**Comment by TimvdLippe** [comment]
@raphw Shall we pick this up soon so that we can prepare for a new major version around September? Given that more and more companies are switching to JDK 17, I think this is the right moment to make the change.

**Comment by raphw** [comment]
Yeah, I think it's hard to estimate the actual impact before it's out so the longer we wait, the more Java 17 is adopted and the more big bang this will get.

**Comment by antch** [comment]
Switching to mockito-inline caused our test suite to take ~2.5x the previous runtime. In reality the non-inline mock-maker is  fine for the vast majority of cases -- is there a potential for a per-use, opt-in mechanism? I'm having a similar issue to #2628 in Java 17, except with `FileOutputStream` but it's far more harmful to upgrade to the inline mock maker than to completely rework the failing test. I'm very worried about eventually having to eat that huge performance hit in order to move forward.

**Comment by raphw** [comment]
You won't have to eat the performance hit because of Mockito, you can certainly pin the subclass mock maker, but the unsafe APIs are going to be deleted and with it, the subclass Mock maker will have to call constructors or break at some point. The JDK will force you to eat the hit, therefore.

Try setting up Surefire to allow self attach, or even add the Byte Buddy agent on the VM startup.

Other than that, the inline mock maker is more expensive due to its design. Nothing we can do about it, we have to go via instrumentation for it.

**Comment by antch** [comment]
Thank you for the response -- when you reference "Surefire self attach" and "Byte Buddy agent", do you mean that those would be used to try to troubleshoot the runtime performance?

**Comment by raphw** [comment]
Yes, it's mentioned in the Mockito javadoc in greater detail.

**Comment by olsavmic** [comment]
Perhaps we could reflect this behavior in the Mockito documentation until the switch to mockito-inline is merged? I can see the inline mock maker mentions final types & methods but not Java modules (https://javadoc.io/doc/org.mockito/mockito-core/latest/org/mockito/Mockito.html#39)

It took me a significant amount of time yesterday to find out why some of our tests were failing as Mockito does not log warning nor fail when it cannot copy internal object state via reflection.

Opening the module or switching to mockito-inline works as expected but the fact that it fails SILENTLY makes finding this issue way more difficult now.

**Comment by TimvdLippe** [comment]
@olsavmic that sounds like a bug to me that we need to fix. Can you file a new issue with a regression test that we can debug?

**Comment by raphw** [comment]
That's always been a weird limitation in Mockito. But it was necessary behavior where some fields could be copied to establish a working spy.

But I agree, the pre-inline behaviour is weird.

**Comment by TimvdLippe** [comment]
I filed #2802 to track our Mockito 5 release. For this issue, we will still need to:

- [ ] Publish a new artifact `mockito-subclass` which is essentially a copy of https://github.com/mockito/mockito/tree/main/subprojects/inline but has https://github.com/mockito/mockito/blob/main/subprojects/inline/src/main/resources/mockito-extensions/org.mockito.plugins.MockMaker updated to specify `mock-maker-subclass` instead per https://github.com/mockito/mockito/blob/0ce902af967cca91cb9ff2beaac95e44406bd175/src/main/java/org/mockito/MockMakers.java#L37 (should be done in Mockito 4)
- [ ] Update https://github.com/mockito/mockito/blob/0ce902af967cca91cb9ff2beaac95e44406bd175/src/main/java/org/mockito/internal/configuration/plugins/DefaultMockitoPlugins.java#L37 to specify `org.mockito.internal.creation.bytebuddy.InlineByteBuddyMockMaker`

I would like to publish the new artifact as part of the last Mockito 4 release, so that users can switch to it and then later make the upgrade to Mockito 5 in which we update our default mockito plugin.

**Comment by TimvdLippe** [comment]
@reta Since you worked on the Java 11 stuff (great work, thanks!), WDYT about picking up the `mockito-subclass` package as well listed in the previous in the comment? That should unblock the rest, so we can start merging your PRs as well 😄

**Comment by reta** [comment]
> @reta Since you worked on the Java 11 stuff (great work, thanks!), WDYT about picking up the `mockito-subclass` package as well listed in the previous in the comment? That should unblock the rest, so we can start merging your PRs as well 😄

@TimvdLippe sure, I will take it :)

**Comment by TimvdLippe** [comment]
4.10.0 has been released with `mockito-subclass`. We can now update our mockmaker in `DefaultMockitoPlugins` and deal with any test failures in our regression tests. There might be some tests that explicitly test subclass mocking, which we need to resolve. If anyone wants to pick that up, let us know!

**Comment by TimvdLippe** [comment]
@reta I know you already did a lot of work here, but would you like to round it off with the `DefaultMockitoPlugins`  change as listed in https://github.com/mockito/mockito/issues/2589#issuecomment-1328262373 ?

**Comment by reta** [comment]
@TimvdLippe sure, happy to help :) Please feel free to assign this issue to me

**Comment by ThrawnCA** [comment]
If the aim is to Just Work out of the box, then the inline mock maker is problematic since it relies on attaching a Java agent (which will _not_ Just Work on JVMs newer than 17).

**Comment by raphw** [comment]
Using the subclass mock maker will neither work out of the box with Unsafe being removed. So we need setup either way.

**Comment by bric3** [comment]
FYI this has been disputed / discussed at length here #3037
